x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DestinationIndex.java

```diff
@@ -220,4 +235,77 @@ private static void checkResultsFieldIsNotPresentInProperties(DataFrameAnalytics
                 DataFrameAnalyticsDest.RESULTS_FIELD.getPreferredName());
         }
     }
+
+    @SuppressWarnings("unchecked")
+    public static Metadata readMetadata(String jobId, MappingMetadata mappingMetadata) {
+        Map<String, Object> mappings = mappingMetadata.getSourceAsMap();
+        Map<String, Object> meta = (Map<String, Object>) mappings.get(META);
+        if (meta == null || DFA_CREATOR.equals(meta.get(CREATED_BY)) == false) {
+            return new NoMetadata();
+        }
+        return new DestMetadata(getVersion(jobId, meta));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static Version getVersion(String jobId, Map<String, Object> meta) {
+        try {
+            Map<String, Object> version = (Map<String, Object>) meta.get(VERSION);
+            String createdVersionString = (String) version.get(CREATED);
+            return Version.fromString(createdVersionString);
+        } catch (Exception e) {
+            logger.error(new ParameterizedMessage("[{}] Could not retrieve destination index version", jobId), e);
+            return null;
+        }
+    }
+
+    public interface Metadata {
```

Comment:
Isn't interface with subclasses an overkill here?
How about turning `Metadata` into a POJO class with 2 fields: `boolean hasMetadata` and `Version version`?

Reply:
I've tried this out and we end up with a single `Metadata` class that looks like this:

```
    private static class Metadata {

        private final boolean hasMetadata;
        private final Version version;

        private Metadata(boolean hasMetadata, @Nullable Version version) {
            this.hasMetadata = hasMetadata;
            this.version = version;
        }

        public boolean hasMetadata() {
            return hasMetadata;
        }

        public boolean isCompatible() {
            if (hasMetadata) {
                return version == null ? false : version.onOrAfter(MIN_COMPATIBLE_VERSION);
            }
            throw new UnsupportedOperationException();
        }

        public String getVersion() {
            if (hasMetadata) {
                return version == null ? "unknown" : version.toString();
            }
            throw new UnsupportedOperationException();
        }
    }
```

While this results to fewer lines of code, I feel the interface approach results into easier to understand and follow code. The `NoMetadata` implementation is straightforward and it frees up the other one not to worry about those unsupported operation exceptions.